demos/mqtt/mqtt_demo_connection_sharing/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,308 @@
+                                 callbackRecordList[ *pCursor ].requiresExactMatch );
+        /* code */
+    } while( ++( *pCursor ) < recordListSize && ( matchFound == false ) );
+
+    return matchFound;
+}
+
+
+/* Design -
+ *   * Common handler can contain logic of processing acks and retries.
+ *   * Only forward PUBLISH message to message dispatcher/handler as individual callbacks
+ *     only want to process incoming PUBLISH message.
+ */
+void SubscriptionManager_DispatchHandler( uint16_t packetIdentifier,
+                                          MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pPublishInfo != NULL );
+    assert( packetIdentifier != MQTT_PACKET_ID_INVALID );
+
+    size_t recordListCursor = 0u;
+
+    while( findRecordForTopic( pPublishInfo->pTopicName,
+                               pPublishInfo->topicNameLength,
+                               &recordListCursor ) != false )
+    {
+        callbackRecordList[ recordListCursor ].callback( packetIdentifier, pPublishInfo );
+    }
+}
+
+bool SubscriptionManager_RegisterCallback( const char * pTopicFilter,
+                                           size_t topicFileterLength,
+                                           SubscriptionManager_Callback_t callback )
+{
+    bool recordAdded = false;
+    size_t availableIndex = 0u;
+
+    /* Search for an available spot in the list to store the record */
+    while( ( callbackRecordList[ availableIndex ].pTopicFilter != NULL ) &&
+           ( availableIndex < recordListSize ) )
+    {
+        availableIndex++;
+    }
+
+    if( availableIndex == recordListSize )
+    {
+        /* Record list is full. */
+    }
+    else
+    {
+        /* Should the topic string be copied? */
+        callbackRecordList[ recordListCount ].pTopicFilter = pTopicFilter;
+        callbackRecordList[ recordListCount ].topicFilterLength = topicFileterLength;
+        callbackRecordList[ recordListCount ].callback = callback;
+
+        recordAdded = true;
+
+        /* Update the max index tracker if a higher index spot is found. */
+        if( availableIndex > maxRecordIndex )
+        {
+            maxRecordIndex = availableIndex;
+        }
+    }
+
+    return recordAdded;
+}
+
+void SubscriptionManager_RemoveCallbackRecord( const char * pTopicFilter,
+                                               size_t topicFileterLength )
+{
+    size_t matchingRecordIndex = 0u;
+    bool recordFound = false;
+    SubscriptionManager_Record_t * pRecord = NULL;
+
+    /* Iterate through the records list to find the matching record. */
+    do
+    {
+        pRecord = &callbackRecordList[ matchingRecordIndex ];
+
+        if( ( topicFileterLength == pRecord->pTopicFilter ) &&
+            ( strncmp( pTopicFilter, pRecord->pTopicFilter, topicFileterLength ) == 0 ) )
+        {
+            recordFound = true;
+        }
+    } while( ( matchingRecordIndex < recordListSize ) && ( recordFound == false ) );
```

Comment:
What if more than one application registers a different callback to the same topic filter? Then this would only remove the first one.

Reply:
Currently, the subscription manager is geared toward(s) a single-threaded demo that will showcase connection sharing for multiple libraries (like Jobs, Shadow) that need different callbacks for processing their topic-specific PUBLISH messages. 

However, even in the case of a multi-threaded use-case, a single application's call to remove its callback shouldn't remove another application's callback for the same topic filter. In that case, the subscription manager would need to store some thread/application-specific information in the records.

Reply:
It is a good point that Muneeb mentioned. What is the purpose of this Subscription manager? Is it only for a single task use case? Is there a plan to improve it to support it for connection sharing between multiple tasks?
If we are extending it to multiple tasks, yea we would need some way to identify the task from which it is called.